Unreadable state store when running a query with a source table with a schema-registry schema

We recently discovered an issue in streams that impacts ksql (https://issues.apache.org/jira/browse/KAFKA-10179). KS can optimize materialization of a topology's source table by using the source topic as the table's change-log rather than creating a new change-log topic. This becomes an issue during restore, because the restore just copies the source topic data into rocksdb. This is a problem for schemas registered with schema-registry. The data in the source topic has a schema id that is registered under the source topic's schema. However, the deserializer for the state store is provided a subject for the store, which doesn't have that schema registered, and deserialization fails.

Streams has proposed a fix in KAFKA-10179. However, the proposed fix - to associate the source topic name with the changelog state store - is not going to work for ksql. In ksql, the schema of the data going into the changelog may be different from the source topic’s schema (e.g. we may only be using some of the source’s fields), and we can’t register that schema with the source topic’s subject in the schema registry. We need to do our own work to solve the following:

How do we want to proceed for new queries? We could use the optimization, but then we must read and materialize the source using the source’s schema, and project out the needed fields in a downstream transformation. Alternatively, we could disable the optimization (provided streams gives us a way of doing this).

How do we want to handle existing queries? Existing queries may be in a corrupted state (e.g. corrupted state stores). How do we handle this? Some options:
- try and detect this state and mark the queries as failed
- detect this from the deserializer and drop corrupted rows
- play some games with the deserializer to look for schemas in both subjects during deserialization

Steps to reproduce:

```sql
create table t_source (id int primary key, `col1` int) with (kafka_topic='t', value_format='avro', partitions=1);
create table t (id int primary key) with (kafka_topic='t', value_format='avro');
create stream s (id int key, col1 int) with (kafka_topic='s', value_format='avro', partitions=1);
insert into t_source  VALUES (1, 1);
insert into s (id, col1) VALUES (1, 1);
SET 'auto.offset.reset'='earliest';
create stream j as select t.*, s.* from s join t on s.col1 = t.id;

--- at this point you need to delete your streams data dir and
--- restart the ksql application

insert into s (id, col1) VALUES (1, 1);
```

The following topics have been created:
```
ksql> SHOW TOPICS;

 Kafka Topic                 | Partitions | Partition Replicas
---------------------------------------------------------------
 J                           | 1          | 1
 default_ksql_processing_log | 1          | 1
 s                           | 1          | 1
 t                           | 1          | 1
---------------------------------------------------------------
```

At this point, schema registry will have the following information (note that `_confluent-ksql-default_query_CSAS_J_0-KafkaTopic_Right-Reduce-changelog-value` is registered in schema registry even though the topic does not exist):

```sh
curl -XGET 'http://localhost:8081/subjects'
["_confluent-ksql-default_query_CSAS_J_0-KafkaTopic_Right-Reduce-changelog-value","t-value","s-value","_confluent-ksql-default_query_CSAS_J_0-Join-repartition-value","J-value"]
```

Importantly, notice the difference between `t-value` and `_confluent-ksql-default_query_CSAS_J_0-KafkaTopic_Right-Reduce-changelog-value`, specifically that the fields are upper-cased in the latter and the id of the two schemas diverge:
```
curl -XGET 'http://localhost:8081/subjects/t-value/versions/1'
{"subject":"...","version":1,"id":1,"schema":"{\"type\":\"record\",\"name\":\"KsqlDataSourceSchema\",\"namespace\":\"io.confluent.ksql.avro_schemas\",\"fields\":[{\"name\":\"col1\",\"type\":[\"null\",\"int\"],\"default\":null}],\"connect.name\":\"io.confluent.ksql.avro_schemas.KsqlDataSourceSchema\"}"}

curl -XGET 'http://localhost:8081/subjects/_confluent-ksql-default_query_CSAS_J_0-Join-repartition-value/versions/1'
{"subject":"...","version":1,"id":4,"schema":"{\"type\":\"record\",\"name\":\"KsqlDataSourceSchema\",\"namespace\":\"io.confluent.ksql.avro_schemas\",\"fields\":[{\"name\":\"COL1\",\"type\":[\"null\",\"int\"],\"default\":null}]}"}
```

The logs will show the following error message:
```
Caused by: org.apache.kafka.common.errors.SerializationException: Error deserializing message from topic: _confluent-ksql-default_query_CSAS_J_0-KafkaTopic_Right-Reduce-changelog
Caused by: org.apache.kafka.connect.errors.DataException: Failed to deserialize data for topic _confluent-ksql-default_query_CSAS_J_0-KafkaTopic_Right-Reduce-changelog to Avro: 
	at io.confluent.connect.avro.AvroConverter.toConnectData(AvroConverter.java:118)
	at io.confluent.ksql.serde.connect.KsqlConnectDeserializer.deserialize(KsqlConnectDeserializer.java:45)
	at io.confluent.ksql.serde.tls.ThreadLocalDeserializer.deserialize(ThreadLocalDeserializer.java:37)
	at io.confluent.ksql.serde.GenericRowSerDe$GenericRowDeserializer.deserialize(GenericRowSerDe.java:304)
	at io.confluent.ksql.serde.GenericRowSerDe$GenericRowDeserializer.deserialize(GenericRowSerDe.java:289)
	at io.confluent.ksql.logging.processing.LoggingDeserializer.deserialize(LoggingDeserializer.java:45)
	at org.apache.kafka.streams.state.internals.ValueAndTimestampDeserializer.deserialize(ValueAndTimestampDeserializer.java:54)
	at org.apache.kafka.streams.state.internals.ValueAndTimestampDeserializer.deserialize(ValueAndTimestampDeserializer.java:27)
	at org.apache.kafka.streams.state.StateSerdes.valueFrom(StateSerdes.java:160)
	at org.apache.kafka.streams.state.internals.MeteredKeyValueStore.outerValue(MeteredKeyValueStore.java:210)
	at org.apache.kafka.streams.state.internals.MeteredKeyValueStore.lambda$get$2(MeteredKeyValueStore.java:133)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:880)
	at org.apache.kafka.streams.state.internals.MeteredKeyValueStore.get(MeteredKeyValueStore.java:133)
	at org.apache.kafka.streams.processor.internals.AbstractReadWriteDecorator$KeyValueStoreReadWriteDecorator.get(AbstractReadWriteDecorator.java:78)
	at org.apache.kafka.streams.kstream.internals.KTableSourceValueGetterSupplier$KTableSourceValueGetter.get(KTableSourceValueGetterSupplier.java:49)
	at org.apache.kafka.streams.kstream.internals.KTableTransformValues$KTableTransformValuesGetter.get(KTableTransformValues.java:144)
	at org.apache.kafka.streams.kstream.internals.KTableTransformValues$KTableTransformValuesGetter.get(KTableTransformValues.java:144)
	at org.apache.kafka.streams.kstream.internals.KStreamKTableJoinProcessor.process(KStreamKTableJoinProcessor.java:77)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.lambda$process$2(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:865)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.process(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:236)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:216)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:168)
	at org.apache.kafka.streams.processor.internals.SourceNode.process(SourceNode.java:100)
	at org.apache.kafka.streams.processor.internals.StreamTask.lambda$process$1(StreamTask.java:716)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:865)
	at org.apache.kafka.streams.processor.internals.StreamTask.process(StreamTask.java:716)
	at org.apache.kafka.streams.processor.internals.TaskManager.process(TaskManager.java:884)
	at org.apache.kafka.streams.processor.internals.StreamThread.runOnce(StreamThread.java:695)
	at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:550)
	at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:509)
Caused by: org.apache.kafka.common.errors.SerializationException: Error retrieving Avro value schema version for id 3
Caused by: io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException: Schema not found; error code: 40403
	at io.confluent.kafka.schemaregistry.client.rest.RestService.sendHttpRequest(RestService.java:292)
	at io.confluent.kafka.schemaregistry.client.rest.RestService.httpRequest(RestService.java:351)
	at io.confluent.kafka.schemaregistry.client.rest.RestService.lookUpSubjectVersion(RestService.java:436)
	at io.confluent.kafka.schemaregistry.client.rest.RestService.lookUpSubjectVersion(RestService.java:423)
	at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getVersionFromRegistry(CachedSchemaRegistryClient.java:228)
	at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getVersion(CachedSchemaRegistryClient.java:356)
	at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer.schemaVersion(AbstractKafkaAvroDeserializer.java:113)
	at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer.deserializeWithSchemaAndVersion(AbstractKafkaAvroDeserializer.java:155)
	at io.confluent.connect.avro.AvroConverter$Deserializer.deserialize(AvroConverter.java:162)
	at io.confluent.connect.avro.AvroConverter.toConnectData(AvroConverter.java:101)
	at io.confluent.ksql.serde.connect.KsqlConnectDeserializer.deserialize(KsqlConnectDeserializer.java:45)
```

Some notes:

* the serializer that serializes the messages into the state store will serialize them under the subject `_confluent-ksql-default_query_CSAS_J_0-KafkaTopic_Right-Reduce-changelog-value` using the schema of the ksql source (_not_ the existing topic schema). 
* if the schema that is being serialized to the state store is exactly the same as the schema of the source topic, then schema registry will reuse the same schemaId (this is why it is necessary to differentiate between `t` and `t_source` in the example above)

I still need to confirm this hypothesis, but I think that for existing queries we should be able to register the exact source schema under the changelog subject so that either (1) the data has the schemaId that was registered in steady state or (2) the data comes from a recovery and it will then be able to still find the correct schemaid in the changelog subject. In the example above, that would mean registering schema id `1` under the `_confluent-ksql-default_query_CSAS_J_0-KafkaTopic_Right-Reduce-changelog-value` subject.

The concern is that this makes it so that the state store in recovery may be larger than the state store in normal operation (because a DDL may be projecting fewer fields than those which are in the topic itself), but I believe this is preferable to data loss.

I'm still thinking about the best way to handle "new queries" that come after the fix for KAFKA-10179 comes in.

--- a/ksqldb-serde/src/main/java/io/confluent/ksql/logging/processing/LoggingDeserializer.java
+++ b/ksqldb-serde/src/main/java/io/confluent/ksql/logging/processing/LoggingDeserializer.java
@@ -17,6 +17,7 @@ package io.confluent.ksql.logging.processing;
 
 import static java.util.Objects.requireNonNull;
 
+import com.google.common.annotations.VisibleForTesting;
 import java.util.Map;
 import java.util.Optional;
 import org.apache.kafka.common.serialization.Deserializer;
@@ -41,11 +42,26 @@ public final class LoggingDeserializer<T> implements Deserializer<T> {
 
   @Override
   public T deserialize(final String topic, final byte[] bytes) {
+    return tryDeserialize(topic, bytes).get();
+  }
+
+  /**
+   * Similar to {@link #deserialize(String, byte[])}, but allows the erroneous case
+   * to delay the log-and-throw behavior until {@link DelayedResult#get()} is called.
+   *
+   * <p>This can be used in the scenarios when an error is expected, such as if a retry
+   * will likely solve the problem, to avoid spamming the processing logger with messages
+   * that are not helpful to the end user.</p>
+   */
+  public DelayedResult<T> tryDeserialize(final String topic, final byte[] bytes) {
     try {
-      return delegate.deserialize(topic, bytes);
+      return new DelayedResult<T>(delegate.deserialize(topic, bytes));
     } catch (final RuntimeException e) {
-      processingLogger.error(new DeserializationError(e, Optional.ofNullable(bytes), topic));
-      throw e;
+      return new DelayedResult<T>(
+          e,
+          new DeserializationError(e, Optional.ofNullable(bytes), topic),
+          processingLogger
+      );
     }
   }
 
@@ -53,4 +69,51 @@ public final class LoggingDeserializer<T> implements Deserializer<T> {
   public void close() {
     delegate.close();
   }
+
+  public static class DelayedResult<T> {
+
+    private final T result;
+    private final RuntimeException error;
+    private final ProcessingLogger processingLogger;
+    private final DeserializationError deserializationError;
+
+    public DelayedResult(
+        final RuntimeException error,
+        final DeserializationError deserializationError,
+        final ProcessingLogger processingLogger
+    ) {
+      this.result = null;
+      this.error = error;
+      this.deserializationError = requireNonNull(deserializationError, "deserializationError");
+      this.processingLogger = requireNonNull(processingLogger, "processingLogger");
+    }
+
+    public DelayedResult(final T result) {
+      this.result = result;
+      this.error = null;
+      this.deserializationError = null;
+      this.processingLogger = null;
+    }
+
+    public boolean isError() {
+      return error != null;
+    }
+
+    @VisibleForTesting
+    RuntimeException getError() {
+      return error;
+    }
+
+    @SuppressWarnings("ConstantConditions")
+    public T get() {
+      if (isError()) {
+        processingLogger.error(deserializationError);
+        throw error;
+      }
+
+      return result;
+    }
+
+  }
+
 }

--- a/ksqldb-serde/src/main/java/io/confluent/ksql/serde/StaticTopicSerde.java
+++ b/ksqldb-serde/src/main/java/io/confluent/ksql/serde/StaticTopicSerde.java
@@ -1,0 +1,152 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.serde;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.confluent.ksql.logging.processing.LoggingDeserializer;
+import io.confluent.ksql.logging.processing.LoggingDeserializer.DelayedResult;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.common.serialization.Serializer;
+
+/**
+ * The {@code StaticTopicSerde} hard codes the topic name that is passed
+ * to the delegate Serde, regardless to what the caller passes in as the
+ * topic. The only exception is that if a deserialization attempt fails,
+ * the deserializer will attempt one more time using the topic that was
+ * passed to the Serde (instead of the hard coded value). In this situation,
+ * the {@code onFailure} callback is called so that the user of this class
+ * can remedy the issue (i.e. register an extra schema under the hard coded
+ * topic). The callback will not be called if both serialization attempts fail.
+ *
+ * <p>This class is intended as a workaround for the issues described in
+ * both KAFKA-10179 and KSQL-5673; specifically, it allows a materialized
+ * state store to use a different topic name than that which Kafka Streams
+ * passes in to the Serde.</p>
+ *
+ * <p><b>Think carefully before reusing this class! It's inteded use case is
+ * very narrow.</b></p>
+ */
+public final class StaticTopicSerde<T> implements Serde<T> {
+
+  public interface Callback {
+
+    /**
+     * This method is called when the {@link Serde#deserializer()}'s produced by
+     * this class' {@link Deserializer#deserialize(String, byte[])} method fails
+     * using the static topic but succeeds using the source topic.
+     *
+     * @param sourceTopic the original topic that was passed in to the deserializer
+     * @param staticTopic the hard coded topic that was passed into the {@code StaticTopicSerde}
+     * @param data        the data that failed deserialization
+     */
+    void onDeserializationFailure(String sourceTopic, String staticTopic, byte[] data);
+  }
+
+  private final Serde<T> delegate;
+  private final String topic;
+  private final Callback onFailure;
+
+  /**
+   * @param topic     the topic to hardcode
+   * @param serde     the delegate serde
+   * @param onFailure a callback to call on failure
+   *
+   * @return a serde which delegates to {@code serde} but passes along {@code topic}
+   *         in place of whatever the actual topic is
+   */
+  public static <S> Serde<S> wrap(
+      final String topic,
+      final Serde<S> serde,
+      final Callback onFailure
+  ) {
+    return new StaticTopicSerde<>(topic, serde, onFailure);
+  }
+
+  private StaticTopicSerde(
+      final String topic,
+      final Serde<T> delegate,
+      final Callback onFailure
+  ) {
+    this.topic = Objects.requireNonNull(topic, "topic");
+    this.delegate = Objects.requireNonNull(delegate, "delegate");
+    this.onFailure = Objects.requireNonNull(onFailure, "onFailure");
+  }
+
+  @Override
+  public void configure(final Map<String, ?> configs, final boolean isKey) {
+    delegate.configure(configs, isKey);
+  }
+
+  @Override
+  public void close() {
+    delegate.close();
+  }
+
+  @Override
+  public Serializer<T> serializer() {
+    final Serializer<T> serializer = delegate.serializer();
+    return (topic, data) -> serializer.serialize(this.topic, data);
+  }
+
+  @Override
+  public Deserializer<T> deserializer() {
+    final Deserializer<T> deserializer = delegate.deserializer();
+
+    if (deserializer instanceof LoggingDeserializer<?>) {
+      final LoggingDeserializer<T> loggingDeserializer = (LoggingDeserializer<T>) deserializer;
+
+      return (topic, data) -> {
+        final DelayedResult<T> staticResult = loggingDeserializer.tryDeserialize(this.topic, data);
+        if (!staticResult.isError()) {
+          return staticResult.get();
+        }
+
+        // if both attempts error, then staticResult.get() will log the error to
+        // the processing log and throw - do not call the callback in this case
+        final DelayedResult<T> sourceResult = loggingDeserializer.tryDeserialize(topic, data);
+        if (sourceResult.isError()) {
+          return staticResult.get();
+        }
+
+        onFailure.onDeserializationFailure(topic, this.topic, data);
+        return sourceResult.get();
+      };
+    }
+
+    return (topic, data) -> {
+      try {
+        return deserializer.deserialize(this.topic, data);
+      } catch (final Exception e) {
+        final T object = deserializer.deserialize(topic, data);
+        onFailure.onDeserializationFailure(topic, this.topic, data);
+        return object;
+      }
+    };
+  }
+
+  @VisibleForTesting
+  public String getTopic() {
+    return topic;
+  }
+
+  @VisibleForTesting
+  public Callback getOnFailure() {
+    return onFailure;
+  }
+}

--- a/ksqldb-serde/src/test/java/io/confluent/ksql/logging/processing/LoggingDeserializerTest.java
+++ b/ksqldb-serde/src/test/java/io/confluent/ksql/logging/processing/LoggingDeserializerTest.java
@@ -16,7 +16,10 @@
 package io.confluent.ksql.logging.processing;
 
 import static io.confluent.ksql.GenericRow.genericRow;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
 import static org.junit.Assert.assertThrows;
+import static org.junit.Assert.assertTrue;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
@@ -24,12 +27,14 @@ import static org.mockito.Mockito.when;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.testing.NullPointerTester;
 import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.logging.processing.LoggingDeserializer.DelayedResult;
 import java.nio.charset.StandardCharsets;
 import java.util.Map;
 import java.util.Optional;
 import java.util.function.Function;
 import org.apache.kafka.common.serialization.Deserializer;
 import org.apache.kafka.connect.data.SchemaAndValue;
+import org.hamcrest.Matchers;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -99,6 +104,20 @@ public class LoggingDeserializerTest {
     verify(delegate).deserialize("some topic", SOME_BYTES);
   }
 
+  @Test
+  public void shouldTryDeserializeWithDelegate() {
+    // Given:
+    when(delegate.deserialize(any(), any())).thenReturn(SOME_ROW);
+
+    // When:
+    final DelayedResult<GenericRow> result = deserializer.tryDeserialize("some topic", SOME_BYTES);
+
+    // Then:
+    verify(delegate).deserialize("some topic", SOME_BYTES);
+    assertThat(result.isError(), is(false));
+    assertThat(result.get(), is(SOME_ROW));
+  }
+
   @Test(expected = ArithmeticException.class)
   public void shouldThrowIfDelegateThrows() {
     // Given:
@@ -125,5 +144,22 @@ public class LoggingDeserializerTest {
 
     // Then:
     verify(processingLogger).error(new DeserializationError(e, Optional.of(SOME_BYTES), "t"));
+  }
+
+  @Test
+  public void shouldDelayLogOnException() {
+    // Given:
+    when(delegate.deserialize(any(), any()))
+        .thenThrow(new RuntimeException("outer",
+            new RuntimeException("inner", new RuntimeException("cause"))));
+
+    // When:
+    final DelayedResult<GenericRow> result = deserializer.tryDeserialize("t", SOME_BYTES);
+
+    // Then:
+    assertTrue(result.isError());
+    assertThrows(RuntimeException.class, result::get);
+    verify(processingLogger)
+        .error(new DeserializationError(result.getError(), Optional.of(SOME_BYTES), "t"));
   }
 }

--- a/ksqldb-serde/src/test/java/io/confluent/ksql/serde/StaticTopicSerdeTest.java
+++ b/ksqldb-serde/src/test/java/io/confluent/ksql/serde/StaticTopicSerdeTest.java
@@ -1,0 +1,166 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.serde;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThrows;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyZeroInteractions;
+import static org.mockito.Mockito.when;
+
+import io.confluent.ksql.logging.processing.DeserializationError;
+import io.confluent.ksql.logging.processing.LoggingDeserializer;
+import io.confluent.ksql.logging.processing.ProcessingLogger;
+import io.confluent.ksql.serde.StaticTopicSerde.Callback;
+import java.util.Optional;
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.common.serialization.Serdes.WrapperSerde;
+import org.apache.kafka.common.serialization.Serializer;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class StaticTopicSerdeTest {
+
+  private static final byte[] SOME_BYTES = new byte[]{1, 2, 3};
+  private static final Object SOME_OBJECT = 1;
+  private static final String STATIC_TOPIC = "static";
+  private static final String SOURCE_TOPIC = "source";
+
+  @Mock
+  private Serializer<Object> delegateS;
+  @Mock
+  private Deserializer<Object> delegateD;
+  @Mock
+  private Callback callback;
+
+  private Serde<Object> staticSerde;
+
+  @Before
+  public void setUp() {
+    final Serde<Object> delegate = new WrapperSerde<>(delegateS, delegateD);
+
+    when(delegateS.serialize(Mockito.any(), Mockito.any())).thenReturn(SOME_BYTES);
+    when(delegateD.deserialize(Mockito.any(), Mockito.any())).thenReturn(SOME_OBJECT);
+
+    staticSerde = StaticTopicSerde.wrap(STATIC_TOPIC, delegate, callback);
+  }
+
+  @Test
+  public void shouldUseDelegateSerializerWithStaticTopic() {
+    // When:
+    final byte[] serialized = staticSerde.serializer().serialize(SOURCE_TOPIC, SOME_OBJECT);
+
+    // Then:
+    verify(delegateS).serialize(STATIC_TOPIC, SOME_OBJECT);
+    assertThat(serialized, is(SOME_BYTES));
+    verifyZeroInteractions(callback);
+  }
+
+  @Test
+  public void shouldUseDelegateDeserializerWithStaticTopic() {
+    // When:
+    final Object deserialized = staticSerde.deserializer().deserialize(SOURCE_TOPIC, SOME_BYTES);
+
+    // Then:
+    verify(delegateD).deserialize(STATIC_TOPIC, SOME_BYTES);
+    assertThat(deserialized, is(SOME_OBJECT));
+    verifyZeroInteractions(callback);
+  }
+
+  @Test
+  public void shouldUseDelegateLoggingDeserializerWithStaticTopic() {
+    // Given:
+    final ProcessingLogger logger = mock(ProcessingLogger.class);
+    final LoggingDeserializer<Object> loggingDelegate = new LoggingDeserializer<>(delegateD, logger);
+    final Serde<Object> delegate = new WrapperSerde<>(delegateS, loggingDelegate);
+
+    staticSerde = StaticTopicSerde.wrap(STATIC_TOPIC, delegate, callback);
+
+    // When:
+    final Object deserialized = staticSerde.deserializer().deserialize(SOURCE_TOPIC, SOME_BYTES);
+
+    // Then:
+    verify(delegateD).deserialize(STATIC_TOPIC, SOME_BYTES);
+    assertThat(deserialized, is(SOME_OBJECT));
+    verifyZeroInteractions(callback);
+    verifyZeroInteractions(logger);
+  }
+
+  @Test
+  public void shouldTrySourceTopicAndCallCallbackOnDeserializationFailure() {
+    // Given:
+    when(delegateD.deserialize(Mockito.eq(STATIC_TOPIC), Mockito.any())).thenThrow(new RuntimeException());
+
+    // When:
+    final Object deserialized = staticSerde.deserializer().deserialize(SOURCE_TOPIC, SOME_BYTES);
+
+    // Then:
+    verify(delegateD).deserialize(STATIC_TOPIC, SOME_BYTES);
+    verify(delegateD).deserialize(SOURCE_TOPIC, SOME_BYTES);
+    verify(callback).onDeserializationFailure(SOURCE_TOPIC, STATIC_TOPIC, SOME_BYTES);
+    assertThat(deserialized, is(SOME_OBJECT));
+  }
+
+  @Test
+  public void shouldTrySourceTopicAndCallCallbackOnDeserializationFailureWithLoggingDeserializer() {
+    // Given:
+    when(delegateD.deserialize(Mockito.eq(STATIC_TOPIC), Mockito.any())).thenThrow(new RuntimeException());
+
+    final ProcessingLogger logger = mock(ProcessingLogger.class);
+    final LoggingDeserializer<Object> loggingDelegate = new LoggingDeserializer<>(delegateD, logger);
+    final Serde<Object> delegate = new WrapperSerde<>(delegateS, loggingDelegate);
+
+    staticSerde = StaticTopicSerde.wrap(STATIC_TOPIC, delegate, callback);
+
+    // When:
+    final Object deserialized = staticSerde.deserializer().deserialize(SOURCE_TOPIC, SOME_BYTES);
+
+    // Then:
+    verifyZeroInteractions(logger);
+    verify(callback).onDeserializationFailure(SOURCE_TOPIC, STATIC_TOPIC, SOME_BYTES);
+    assertThat(deserialized, is(SOME_OBJECT));
+  }
+
+  @Test
+  public void shouldLogOriginalFailureIfBothFail() {
+    // Given:
+    when(delegateD.deserialize(Mockito.any(), Mockito.any())).thenThrow(new RuntimeException());
+
+    final ProcessingLogger logger = mock(ProcessingLogger.class);
+    final LoggingDeserializer<Object> loggingDelegate = new LoggingDeserializer<>(delegateD, logger);
+    final Serde<Object> delegate = new WrapperSerde<>(delegateS, loggingDelegate);
+
+    staticSerde = StaticTopicSerde.wrap(STATIC_TOPIC, delegate, callback);
+
+    // When:
+    final RuntimeException err = assertThrows(
+        RuntimeException.class,
+        () -> staticSerde.deserializer().deserialize(SOURCE_TOPIC, SOME_BYTES));
+
+    // Then:
+    verify(logger).error(new DeserializationError(err, Optional.of(SOME_BYTES), STATIC_TOPIC));
+    verifyZeroInteractions(callback);
+  }
+
+}

--- a/ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/RegisterSchemaCallback.java
+++ b/ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/RegisterSchemaCallback.java
@@ -1,0 +1,60 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.streams;
+
+import io.confluent.kafka.schemaregistry.ParsedSchema;
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+import io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException;
+import io.confluent.ksql.serde.StaticTopicSerde;
+import io.confluent.ksql.util.KsqlConstants;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class RegisterSchemaCallback implements StaticTopicSerde.Callback {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RegisterSchemaCallback.class);
+  private final SchemaRegistryClient srClient;
+
+  RegisterSchemaCallback(final SchemaRegistryClient srClient) {
+    this.srClient = Objects.requireNonNull(srClient, "srClient");
+  }
+
+  @Override
+  public void onDeserializationFailure(
+      final String source,
+      final String changelog,
+      final byte[] data
+  ) {
+    final String sourceSubject = source + KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX;
+    final String changelogSubject = changelog + KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX;
+    try {
+      // all schema registry events start with a magic byte 0x0 and then four bytes
+      // indicating the schema id - we extract that schema id from the data that failed
+      // to deserialize and then register it into the changelog subject
+      final int id = ByteBuffer.wrap(data, 1, Integer.BYTES).getInt();
+
+      LOG.info("Trying to fetch & register schema id {} under subject {}", id, changelogSubject);
+      final ParsedSchema schema = srClient.getSchemaBySubjectAndId(sourceSubject, id);
+      srClient.register(changelogSubject, schema);
+    } catch (IOException | RestClientException e) {
+      LOG.warn("Failed during deserialization callback for topic " + source, e);
+    }
+  }
+
+}

--- a/ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/SourceBuilder.java
+++ b/ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/SourceBuilder.java
@@ -14,6 +14,8 @@
 
 package io.confluent.ksql.execution.streams;
 
+import static io.confluent.ksql.util.KsqlConfig.KSQL_PERSISTENT_QUERY_NAME_PREFIX_CONFIG;
+import static io.confluent.ksql.util.KsqlConfig.KSQL_SERVICE_ID_CONFIG;
 import static java.util.Objects.requireNonNull;
 
 import com.google.common.collect.ImmutableList;
@@ -36,8 +38,13 @@ import io.confluent.ksql.execution.timestamp.TimestampColumn;
 import io.confluent.ksql.schema.ksql.Column;
 import io.confluent.ksql.schema.ksql.LogicalSchema;
 import io.confluent.ksql.schema.ksql.PhysicalSchema;
+import io.confluent.ksql.serde.FormatFactory;
+import io.confluent.ksql.serde.FormatInfo;
+import io.confluent.ksql.serde.StaticTopicSerde;
+import io.confluent.ksql.serde.StaticTopicSerde.Callback;
 import io.confluent.ksql.serde.WindowInfo;
 import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.ReservedInternalTopics;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
@@ -63,9 +70,12 @@ import org.apache.kafka.streams.kstream.Windowed;
 import org.apache.kafka.streams.processor.ProcessorContext;
 import org.apache.kafka.streams.processor.TimestampExtractor;
 import org.apache.kafka.streams.state.KeyValueStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public final class SourceBuilder {
 
+  private static final Logger LOG = LoggerFactory.getLogger(SourceBuilder.class);
   private static final Collection<?> NULL_WINDOWED_KEY_COLUMNS = Collections.unmodifiableList(
       Arrays.asList(null, null, null)
   );
@@ -176,11 +186,12 @@ public final class SourceBuilder {
         consumedFactory
     );
 
+    final String stateStoreName = tableChangeLogOpName(source.getProperties());
     final Materialized<Struct, GenericRow, KeyValueStore<Bytes, byte[]>> materialized =
         materializedFactory.create(
             keySerde,
             valueSerde,
-            tableChangeLogOpName(source.getProperties())
+            stateStoreName
         );
 
     final KTable<Struct, GenericRow> ktable = buildKTable(
@@ -188,7 +199,9 @@ public final class SourceBuilder {
         queryBuilder,
         consumed,
         nonWindowedKeyGenerator(source.getSourceSchema()),
-        materialized
+        materialized,
+        valueSerde,
+        stateStoreName
     );
 
     return KTableHolder.unmaterialized(
@@ -225,11 +238,12 @@ public final class SourceBuilder {
         consumedFactory
     );
 
+    final String stateStoreName = tableChangeLogOpName(source.getProperties());
     final Materialized<Windowed<Struct>, GenericRow, KeyValueStore<Bytes, byte[]>> materialized =
         materializedFactory.create(
             keySerde,
             valueSerde,
-            tableChangeLogOpName(source.getProperties())
+            stateStoreName
         );
 
     final KTable<Windowed<Struct>, GenericRow> ktable = buildKTable(
@@ -237,7 +251,9 @@ public final class SourceBuilder {
         queryBuilder,
         consumed,
         windowedKeyGenerator(source.getSourceSchema()),
-        materialized
+        materialized,
+        valueSerde,
+        stateStoreName
     );
 
     return KTableHolder.unmaterialized(
@@ -292,16 +308,26 @@ public final class SourceBuilder {
       final KsqlQueryBuilder queryBuilder,
       final Consumed<K, GenericRow> consumed,
       final Function<K, Collection<?>> keyGenerator,
-      final Materialized<K, GenericRow, KeyValueStore<Bytes, byte[]>> materialized
+      final Materialized<K, GenericRow, KeyValueStore<Bytes, byte[]>> materialized,
+      final Serde<GenericRow> valueSerde,
+      final String stateStoreName
   ) {
     final boolean forceChangelog = streamSource instanceof TableSource
         && ((TableSource) streamSource).isForceChangelog();
 
     final KTable<K, GenericRow> table;
     if (!forceChangelog) {
+      final String changelogTopic = changelogTopic(queryBuilder, stateStoreName);
+      final Callback onFailure = getRegisterCallback(
+          queryBuilder, streamSource.getFormats().getValueFormat());
+
       table = queryBuilder
           .getStreamsBuilder()
-          .table(streamSource.getTopicName(), consumed, materialized);
+          .table(
+              streamSource.getTopicName(),
+              consumed.withValueSerde(StaticTopicSerde.wrap(changelogTopic, valueSerde, onFailure)),
+              materialized
+          );
     } else {
       final KTable<K, GenericRow> source = queryBuilder
           .getStreamsBuilder()
@@ -316,6 +342,46 @@ public final class SourceBuilder {
 
     return table
         .transformValues(new AddKeyAndTimestampColumns<>(keyGenerator));
+  }
+
+  private static StaticTopicSerde.Callback getRegisterCallback(
+      final KsqlQueryBuilder builder,
+      final FormatInfo valueFormat
+  ) {
+    final boolean schemaRegistryEnabled = !builder
+        .getKsqlConfig()
+        .getString(KsqlConfig.SCHEMA_REGISTRY_URL_PROPERTY)
+        .isEmpty();
+
+    final boolean useSR = FormatFactory
+        .fromName(valueFormat.getFormat())
+        .supportsSchemaInference();
+
+    if (!schemaRegistryEnabled || !useSR) {
+      return (t1, t2, data) -> { };
+    }
+
+    return new RegisterSchemaCallback(builder.getServiceContext().getSchemaRegistryClient());
+  }
+
+  /**
+   * This code mirrors the logic that generates the name for changelog topics
+   * in kafka streams, which follows the pattern:
+   * <pre>
+   *    applicationID + "-" + stateStoreName + "-changelog".
+   * </pre>
+   */
+  private static String changelogTopic(
+      final KsqlQueryBuilder queryBuilder,
+      final String stateStoreName
+  ) {
+    return ReservedInternalTopics.KSQL_INTERNAL_TOPIC_PREFIX
+        + queryBuilder.getKsqlConfig().getString(KSQL_SERVICE_ID_CONFIG)
+        + queryBuilder.getKsqlConfig().getString(KSQL_PERSISTENT_QUERY_NAME_PREFIX_CONFIG)
+        + queryBuilder.getQueryId().toString()
+        + "-"
+        + stateStoreName
+        + "-changelog";
   }
 
   private static TimestampExtractor timestampExtractor(

--- a/ksqldb-streams/src/test/java/io/confluent/ksql/execution/streams/RegisterSchemaCallbackTest.java
+++ b/ksqldb-streams/src/test/java/io/confluent/ksql/execution/streams/RegisterSchemaCallbackTest.java
@@ -1,0 +1,58 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.streams;
+
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import io.confluent.kafka.schemaregistry.ParsedSchema;
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+import io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException;
+import io.confluent.ksql.util.KsqlConstants;
+import java.io.IOException;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class RegisterSchemaCallbackTest {
+
+  private static final String SUFFIX = KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX;
+  private static final String SOURCE = "s1";
+  private static final String CHANGELOG = "s2";
+  private static final int ID = 1;
+  private static final byte[] SOME_DATA = new byte[]{0x0, 0x0, 0x0, 0x0, 0x1};
+
+  @Mock
+  private SchemaRegistryClient srClient;
+  @Mock
+  private ParsedSchema schema;
+  
+  @Test
+  public void shouldRegisterIdFromData() throws IOException, RestClientException {
+    // Given:
+    when(srClient.getSchemaBySubjectAndId(SOURCE + SUFFIX, ID)).thenReturn(schema);
+    final RegisterSchemaCallback call = new RegisterSchemaCallback(srClient);
+
+    // When:
+    call.onDeserializationFailure(SOURCE, CHANGELOG, SOME_DATA);
+
+    // Then:
+    verify(srClient).register(CHANGELOG + SUFFIX, schema);
+  }
+
+}

--- a/ksqldb-streams/src/test/java/io/confluent/ksql/execution/streams/SourceBuilderTest.java
+++ b/ksqldb-streams/src/test/java/io/confluent/ksql/execution/streams/SourceBuilderTest.java
@@ -18,8 +18,10 @@ package io.confluent.ksql.execution.streams;
 import static io.confluent.ksql.GenericRow.genericRow;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.is;
 import static org.junit.Assert.assertThrows;
+import static org.junit.Assert.assertTrue;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyString;
 import static org.mockito.ArgumentMatchers.eq;
@@ -31,6 +33,7 @@ import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
 import com.google.common.collect.ImmutableMap;
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
 import io.confluent.ksql.GenericRow;
 import io.confluent.ksql.execution.builder.KsqlQueryBuilder;
 import io.confluent.ksql.execution.context.QueryContext;
@@ -48,12 +51,16 @@ import io.confluent.ksql.execution.plan.WindowedTableSource;
 import io.confluent.ksql.execution.timestamp.TimestampColumn;
 import io.confluent.ksql.logging.processing.ProcessingLogger;
 import io.confluent.ksql.name.ColumnName;
+import io.confluent.ksql.query.QueryId;
 import io.confluent.ksql.schema.ksql.LogicalSchema;
 import io.confluent.ksql.schema.ksql.PhysicalSchema;
 import io.confluent.ksql.schema.ksql.types.SqlTypes;
+import io.confluent.ksql.serde.FormatFactory;
 import io.confluent.ksql.serde.FormatInfo;
 import io.confluent.ksql.serde.SerdeOption;
+import io.confluent.ksql.serde.StaticTopicSerde;
 import io.confluent.ksql.serde.WindowInfo;
+import io.confluent.ksql.services.ServiceContext;
 import io.confluent.ksql.util.KsqlConfig;
 import java.util.HashSet;
 import java.util.Optional;
@@ -170,10 +177,16 @@ public class SourceBuilderTest {
   private Materialized<Object, GenericRow, KeyValueStore<Bytes, byte[]>> materialized;
   @Mock
   private ProcessingLogger processingLogger;
+  @Mock
+  private ServiceContext serviceContext;
+  @Mock
+  private SchemaRegistryClient srClient;
   @Captor
   private ArgumentCaptor<ValueTransformerWithKeySupplier<?, GenericRow, GenericRow>> transformSupplierCaptor;
   @Captor
   private ArgumentCaptor<TimestampExtractor> timestampExtractorCaptor;
+  @Captor
+  private ArgumentCaptor<StaticTopicSerde<GenericRow>> serdeCaptor;
   private final GenericRow row = genericRow("baz", 123);
   private PlanBuilder planBuilder;
 
@@ -199,12 +212,16 @@ public class SourceBuilderTest {
     when(kTable.transformValues(any(ValueTransformerWithKeySupplier.class))).thenReturn(kTable);
     when(queryBuilder.buildKeySerde(any(), any(), any())).thenReturn(keySerde);
     when(queryBuilder.buildValueSerde(any(), any(), any())).thenReturn(valueSerde);
+    when(queryBuilder.getServiceContext()).thenReturn(serviceContext);
+    when(queryBuilder.getQueryId()).thenReturn(new QueryId("id"));
     when(queryBuilder.getKsqlConfig()).thenReturn(KSQL_CONFIG);
     when(processorCtx.timestamp()).thenReturn(A_ROWTIME);
+    when(serviceContext.getSchemaRegistryClient()).thenReturn(srClient);
     when(streamsFactories.getConsumedFactory()).thenReturn(consumedFactory);
     when(streamsFactories.getMaterializedFactory()).thenReturn(materializationFactory);
     when(materializationFactory.create(any(), any(), any()))
         .thenReturn((Materialized) materialized);
+    when(valueFormatInfo.getFormat()).thenReturn(FormatFactory.AVRO.name());
 
     planBuilder = new KSPlanBuilder(
         queryBuilder,
@@ -315,6 +332,7 @@ public class SourceBuilderTest {
   public void shouldApplyCorrectTransformationsToSourceTableWithoutForcingChangelog() {
     // Given:
     givenUnwindowedSourceTable(false);
+    when(consumed.withValueSerde(any())).thenReturn(consumed);
 
     // When:
     final KTableHolder<Struct> builtKTable = tableSource.build(planBuilder);
@@ -329,6 +347,28 @@ public class SourceBuilderTest {
     verify(consumedFactory).create(keySerde, valueSerde);
     verify(consumed).withTimestampExtractor(any());
     verify(consumed).withOffsetResetPolicy(AutoOffsetReset.EARLIEST);
+
+    verify(consumed).withValueSerde(serdeCaptor.capture());
+    final StaticTopicSerde<GenericRow> value = serdeCaptor.getValue();
+    assertThat(value.getTopic(), is("_confluent-ksql-default_query_id-base-Reduce-changelog"));
+  }
+
+  @Test
+  public void shouldApplyCreateSchemaRegistryCallbackIfSchemaRegistryIsEnabled() {
+    // Given:
+    when(queryBuilder.getKsqlConfig()).thenReturn(
+        KSQL_CONFIG.cloneWithPropertyOverwrite(
+            ImmutableMap.of(KsqlConfig.SCHEMA_REGISTRY_URL_PROPERTY, "foo")));
+    givenUnwindowedSourceTable(false);
+    when(consumed.withValueSerde(any())).thenReturn(consumed);
+
+    // When:
+    tableSource.build(planBuilder);
+
+    // Then:
+    verify(consumed).withValueSerde(serdeCaptor.capture());
+    final StaticTopicSerde<GenericRow> value = serdeCaptor.getValue();
+    assertThat(value.getOnFailure(), instanceOf(RegisterSchemaCallback.class));
   }
 
 

